### PR TITLE
[terraform-resources] refactor desired state and vault handling

### PR DESCRIPTION
**Pull request by geoberle**
## Goal
Refactor the desired state and vault handling for terraform resource output. This change introduces additional unittesting and typehints, to prepare for future changes like terraform output formats and shared terraform resources.

_Scope:_ refactor the the use of terraform resources from qontrac-server within terraform-resources and terraform-client.
_Out of scope:_ terrascript-client was not touched to keep the size of this change manageable and reviewable. A followup change will start using the changes of this MR.

## How to review
Please review commit by commit. Each commit contains a smaller logical change along with a detailed commit message and respective type hints and unittests.

### TerraformResourceSpec dataclass
This refactor introduces a dataclass `TerraformResourceSpec` that holds data and context for terraform resources, namely:
* the raw resource specification from qontract-server
* the raw namespace and cluster data from qontract-server
* the output data from the last terraform run
this way a resource spec can be passed around with its close context. also frequently used fields are defined as properties with typehints.

### Desired state management
The population of the `ResourceInventory` and creation of the kube `Secret` have been extracted from `terraform_client.py` to `TerraformResourceSpec`. This way the terraform client can concentrate exclusively on terraform matters and is not concerned with close kubernetes matters like `Secret` management. terraform-client keeps the responsibility of parsing terraform output, and making the per-resource grouped outputs available as the `secret` field of the `TerraformResourceSpec` - see `populate_terraform_output_secrets`

This change also makes sense since the required context to build a `Secret` is now available within `TerraformResourceSpec`. Also, implementation of [terraform output formats](https://github.com/app-sre/qontract-schemas/pull/119) will be contained within the `TerraformResourceSpec`.

### Use TerraformResourceSpec as source for Vault population
The secret embedded in `TerraformResourceSpec` is used to write the terraform out data to vault. Previously the `ResourceInventory` was used for that. Upcoming changes will introduce [terraform output formats](https://github.com/app-sre/qontract-schemas/pull/119), that will change the way the kuberrnetes secret resources are rendered, and therefore the value in the `ResourceInventory` will not necessary the format anymore that needs to be written to Vault.

### Typehints & Tests
Typehints and tests were introduced where code was changed or where it was necessary for mypy to be happy.

## How this change was tested

* dry running for all major accounts with the code before and after the change resulted in the same terraform files
* added extensive tests to the central functions that were touched
* comparing desired and current state for kubernetes secrets with production data
* comparing desired and current state for vault data with production data

### CODE REVIEW

File: reconcile/terraform_resources.py [hunk 1]
```diff
@@ -443,12 +460,22 @@ def setup(dry_run, print_to_file, thread_pool_size, internal,
...
-                                         settings=settings)
+
+    # build the resource specs
+    resource_specs = init_tf_resource_specs(tf_namespaces, account_name)
+
+    # initialize terrascript (scripting engine to generate terraform manifests)
+    ts, working_dirs = init_working_dirs(accounts, thread_pool_size, settings=settings)
+
+    # initialize terraform client
+    # it used to plan and apply according to the output of terrascript
```
Comment (rporres) [inline reconcile/terraform_resources.py:478]:
```suggestion
    # it is used to plan and apply according to the output of terrascript
```

File: reconcile/terraform_resources.py [hunk 2]
```diff
@@ -549,7 +627,13 @@ def run(dry_run, print_to_file=None,
         if err:
             cleanup_and_exit(tf, err)
 
-    tf.populate_desired_state(ri, oc_map, tf_namespaces, account_name)
+    # refresh output data after terraform apply
+    tf.populate_terraform_output_secrets(
+        resource_specs=resource_specs,
+        init_rds_replica_source=True
+    )
+    # populate resourceinventory with latest output data
```
Comment (rporres) [inline reconcile/terraform_resources.py:635]:
```suggestion
    # populate the resource inventory with latest output data
```

File: reconcile/utils/terraform_resource_spec.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        annotation_str = self.resource.get("annotations")
+        if annotation_str:
+            return json.loads(annotation_str)
+        else:
+            return {}
+
+    def get_secret_field(self, field: str) -> Optional[str]:
+        return self.secret.get(field)
+
+    def id_object(self) -> "TerraformResourceIdentifier":
```
Comment (rporres) [inline reconcile/utils/terraform_resource_spec.py:54]: why the stringified return type hint?

File: reconcile/utils/terraform_resource_spec.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+                "dict does not include required both keys 'identifier' and 'provider'"
+            )
+        return TerraformResourceIdentifier(
+            identifier=cast(str, data["identifier"]),
+            provider=cast(str, data["provider"]),
+            account=cast(str, data["account"]),
+        )
+
+    @staticmethod
+    def from_output_prefix(
```
Comment (rporres) [inline reconcile/utils/terraform_resource_spec.py:112]: what is this used for?

File: reconcile/utils/terraform_resource_spec.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        else:
+            return {}
+
+    def get_secret_field(self, field: str) -> Optional[str]:
+        return self.secret.get(field)
+
+    def id_object(self) -> "TerraformResourceIdentifier":
+        return TerraformResourceIdentifier.from_dict(self.resource)
+
+    def construct_oc_resource(self, integration: str, integration_version: str) -> OR:
```
Comment (rporres) [inline reconcile/utils/terraform_resource_spec.py:57]: if this is building a secret, why not something like?
```suggestion
    def construct_oc_secret(self, integration: str, integration_version: str) -> OR:
```
or even using the more common `build`
```suggestion
    def build_oc_secret(self, integration: str, integration_version: str) -> OR:
```

**Review by rporres — COMMENTED** [review]
Still chewing this, it looks great. I like a lot the separation between tf and oc and the use of dataclasses

File: reconcile/utils/terraform_resource_spec.py (hunk 5 above)
Comment (geoberle) [reply reconcile/utils/terraform_resource_spec.py]: i used the second name suggestion - thx! [b023c96](https://github.com/app-sre/qontract-reconcile/pull/2331/commits/b023c965c1dced92eb46ce03f7e2cb1343cfe509)

File: reconcile/utils/terraform_resource_spec.py (hunk 4 above)
Comment (geoberle) [reply reconcile/utils/terraform_resource_spec.py]: a leftover from some experiments. i removed it in [b023c96](https://github.com/app-sre/qontract-reconcile/pull/2331/commits/b023c965c1dced92eb46ce03f7e2cb1343cfe509)

File: reconcile/utils/terraform_resource_spec.py (hunk 3 above)
Comment (geoberle) [reply reconcile/utils/terraform_resource_spec.py:54]: that is a forward reference because the type is later defined in the file

File: reconcile/terraform_resources.py (hunk 2 above)
Comment (geoberle) [reply reconcile/terraform_resources.py]: [b023c96](https://github.com/app-sre/qontract-reconcile/pull/2331/commits/b023c965c1dced92eb46ce03f7e2cb1343cfe509)

File: reconcile/terraform_resources.py (hunk 1 above)
Comment (geoberle) [reply reconcile/terraform_resources.py]: [b023c96](https://github.com/app-sre/qontract-reconcile/pull/2331/commits/b023c965c1dced92eb46ce03f7e2cb1343cfe509)

**Comment by geoberle** [comment]
> Still chewing this
i know, it is a big one. thanks for taking the efforts to dig thought this ❤️ 
followup refactors will be smaller because of the heavy lifting from this one!

File: reconcile/utils/terraform_resource_spec.py (hunk 3 above)
Comment (rporres) [reply reconcile/utils/terraform_resource_spec.py:54]: ok! tjx

File: reconcile/terraform_resources.py [hunk 6]
```diff
@@ -432,8 +437,20 @@ def init_working_dirs(accounts: list[dict[str, Any]],
...
-def setup(dry_run, print_to_file, thread_pool_size, internal,
-          use_jump_host, account_name, extra_labels):
+def setup(
+    dry_run: bool,
+    print_to_file: str,
+    thread_pool_size: int,
+    internal: str,
+    use_jump_host: bool,
+    account_name: Optional[str],
+    extra_labels,
```
Comment (steveteahan) [inline reconcile/terraform_resources.py:447]: Any reason not to type `extra_labels`?

File: reconcile/terraform_resources.py [hunk 7]
```diff
@@ -469,10 +496,12 @@ def setup(dry_run, print_to_file, thread_pool_size, internal,
                           ocm_map=ocm_map)
     ts.dump(print_to_file, existing_dirs=working_dirs)
 
-    return ri, oc_map, tf, tf_namespaces
+    return ri, oc_map, tf, resource_specs
 
 
-def filter_tf_namespaces(namespaces, account_name):
+def filter_tf_namespaces(
+    namespaces: list[dict[str, Any]], account_name: Optional[str]
```
Comment (steveteahan) [inline reconcile/terraform_resources.py:503]: Can we use `Iterable` and `Mapping` instead for the input types, or is it not possible here? There are a few other instances of this in the PR, but I'll avoid on commenting on them all.

File: reconcile/terraform_resources.py [hunk 8]
```diff
@@ -502,27 +549,58 @@ def cleanup_and_exit(tf=None, status=False, working_dirs=None):
     sys.exit(status)
 
 
-def write_outputs_to_vault(vault_path, ri):
+def write_outputs_to_vault(vault_path: str, resoure_specs: TerraformResourceSpecDict) -> None:
```
Comment (steveteahan) [inline reconcile/terraform_resources.py:552]: typo: resoure_specs -> resource_specs ?

File: reconcile/utils/terraform_resource_spec.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+    @property
+    def output_prefix(self) -> str:
+        return f"{self.identifier}-{self.provider}"
+
+    @staticmethod
+    def from_dict(data: dict[str, Any]) -> "TerraformResourceIdentifier":
+        if "identifier" not in data or "provider" not in data or "account" not in data:
+            raise ValueError(
+                "dict does not include required both keys 'identifier' and 'provider'"
```
Comment (steveteahan) [inline reconcile/utils/terraform_resource_spec.py:103]: This seems to be missing `account`? Maybe a good sign that we should have a variable here and just change this to: `dict is missing one of the required fields: {required_fields}`?

**Review by steveteahan — CHANGES_REQUESTED** [review]
Very nice descriptions on the new tests. Thanks for that! Some high level feedback to start, but I'll loop back in the morning to take another look.

**Comment by steveteahan** [comment]
I guess the other thing I'm interested is what manual testing was done? I'm sure you've done this, but it just might be good to list the scenarios explicitly for the benefit of others.

File: reconcile/utils/terraform_client.py [hunk 10]
```diff
@@ -511,38 +457,51 @@ def format_output(self, output, type):
...
-            "type": "Opaque",
-            "metadata": {
-                "name": name,
-                "annotations": {
-                    "qontract.recycle": "true"
-                }
-            },
-            "data": {}
-        }
+    def populate_terraform_output_secrets(self, resource_specs: TerraformResourceSpecDict, init_rds_replica_source: bool = False) -> None:
```
Comment (Piojo) [inline reconcile/utils/terraform_client.py:460]: How come pylint or Black arent whining about the length of this 137-character line?

File: reconcile/utils/terraform_client.py [hunk 11]
```diff
@@ -511,38 +457,51 @@ def format_output(self, output, type):
...
 
-        if annotations:
-            anno_dict = json.loads(base64.b64decode(annotations))
-            body["metadata"]["annotations"].update(anno_dict)
+        self._populate_terraform_output_secrets(resource_specs, existing_secets, self.integration_prefix, replicas_info)
 
-        for k, v in data.items():
-            if self.integration_prefix in k:
+    @staticmethod
+    def _populate_terraform_output_secrets(resource_specs: TerraformResourceSpecDict, existing_secrets: dict[str, dict[str, dict[str, str]]], integration_prefix: str, replica_sources: dict[str, dict[str, str]]) -> None:
```
Comment (Piojo) [inline reconcile/utils/terraform_client.py:476]: This line is definitely too long.

File: reconcile/utils/terraform_resource_spec.py [hunk 12]
```diff
@@ -1,6 +1,8 @@
 from dataclasses import dataclass, field
 import json
 from typing import Any, Optional, cast
+import base64
+from reconcile.utils.openshift_resource import OpenshiftResource as OR
```
Comment (Piojo) [inline reconcile/utils/terraform_resource_spec.py:5]: Can we stop the pattern of calling `OpenshiftResource` "`OR`"? Short identifiers should be used in local variables only.

File: reconcile/utils/terraform_resource_spec.py [hunk 13]
```diff
@@ -49,6 +51,39 @@ def annotations(self) -> dict[str, str]:
     def get_secret_field(self, field: str) -> Optional[str]:
         return self.secret.get(field)
 
+    def id_object(self) -> "TerraformResourceIdentifier":
+        return TerraformResourceIdentifier.from_dict(self.resource)
+
+    def construct_oc_resource(self, integration: str, integration_version: str) -> OR:
```
Comment (Piojo) [inline reconcile/utils/terraform_resource_spec.py]: ... because I'm banner-blind to import statements, and when I see a boolean `OR` here I get confused (did I mention it's late?).

File: reconcile/utils/terraform_client.py [hunk 14]
```diff
@@ -331,120 +331,66 @@ def terraform_apply(self, apply_spec):
...
+
+                # Also, we look for replicate_source_db in the overrides
+                override_replica_src_db = None
+                overrides = tf_resource.get('overrides')
+                if overrides is not None:
+                    override_replica_src_db = json.loads(overrides).get(
+                        'replicate_source_db'
+                    )
+                if override_replica_src_db is not None:
+                    replica_src_db = override_replica_src_db
```
[the remark this answers is not in the document]
Comment (geoberle) [reply reconcile/utils/terraform_client.py:378]: i did not want to modify code that was not logically touched but you are right, this one deserves touching

File: reconcile/terraform_resources.py (hunk 6 above)
Comment (geoberle) [reply reconcile/terraform_resources.py]: fixed

File: reconcile/terraform_resources.py (hunk 7 above)
Comment (geoberle) [reply reconcile/terraform_resources.py]: fixed

File: reconcile/terraform_resources.py (hunk 8 above)
Comment (geoberle) [reply reconcile/terraform_resources.py]: fixed

File: reconcile/utils/terraform_resource_spec.py (hunk 9 above)
Comment (geoberle) [reply reconcile/utils/terraform_resource_spec.py]: i changed it to rely on the pydantic dataclass to complaint about missing stuff

File: setup.py [hunk 15]
```diff
@@ -2,7 +2,7 @@
 
 setup(
     name="qontract-reconcile",
-    version="0.5.3",
+    version="0.5.4",
```
[the remark this answers is not in the document]
Comment (geoberle) [reply setup.py]: 😭

File: reconcile/utils/terraform_resource_spec.py [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+    account: str
+
+    @property
+    def output_prefix(self) -> str:
+        return f"{self.identifier}-{self.provider}"
+
+    @staticmethod
+    def from_dict(data: dict[str, Any]) -> "TerraformResourceIdentifier":
+        if "identifier" not in data or "provider" not in data or "account" not in data:
+            raise ValueError(
```
[the remark this answers is not in the document]
Comment (geoberle) [reply reconcile/utils/terraform_resource_spec.py]: i changed it to rely on the dataclass to complaint so it raises a pydantic ValidationError

File: reconcile/test/test_utils_terraform_resource_spec.py [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+
+
+def test_identifier_account_missing():
+    with pytest.raises(ValueError):
+        TRI.from_dict({"identifier": "i", "provider": "p"})
+
+
+def test_identifier_provider_missing():
+    with pytest.raises(ValueError):
+        TRI.from_dict({"identifier": "i", "account": "a"})
```
[the remark this answers is not in the document]
Comment (geoberle) [reply reconcile/test/test_utils_terraform_resource_spec.py]: fixed

File: reconcile/terraform_resources.py [hunk 18]
```diff
@@ -514,15 +561,30 @@ def write_outputs_to_vault(vault_path, ri):
...
+    internal=None,
+    use_jump_host=True,
+    light=False,
+    vault_output_path="",
+    account_name=None,
+    extra_labels=None,
+    defer=None,
+):
+
+    ri, oc_map, tf, tf_namespaces, resource_specs = setup(
```
[the remark this answers is not in the document]
Comment (geoberle) [reply reconcile/terraform_resources.py]: sure we can. in the next refactor wave i'm going to address this

File: reconcile/utils/terraform_client.py [hunk 19]
```diff
@@ -331,120 +331,66 @@ def terraform_apply(self, apply_spec):
...
+            # that have a replica_source defined in app-interface
+            replica_src = tf_resource.get('replica_source')
+
+            if replica_src is None:
+                # When replica_source is not there, we look for
+                # replicate_source_db in the defaults
+                replica_src_db = None
+                defaults_ref = tf_resource.get('defaults')
+                if defaults_ref is not None:
+                    defaults_res = gql.get_resource(
```
[the remark this answers is not in the document]
Comment (geoberle) [reply reconcile/utils/terraform_client.py]: fixed

File: reconcile/utils/terraform_client.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (geoberle) [reply reconcile/utils/terraform_client.py]: done

File: reconcile/utils/terraform_client.py (hunk 11 above)
Comment (geoberle) [reply reconcile/utils/terraform_client.py]: fixed

File: reconcile/utils/terraform_resource_spec.py (hunk 12 above)
Comment (geoberle) [reply reconcile/utils/terraform_resource_spec.py]: done

File: reconcile/utils/terraform_resource_spec.py (hunk 13 above)
Comment (geoberle) [reply reconcile/utils/terraform_resource_spec.py]: fixed in this file

File: reconcile/utils/terraform_client.py (hunk 10 above)
Comment (geoberle) [reply reconcile/utils/terraform_client.py]: i shortened the line

**Comment by geoberle** [comment]
> I guess the other thing I'm interested is what manual testing was done? I'm sure you've done this, but it just might be good to list the scenarios explicitly for the benefit of others.

i was doing the following
* i was running in dry run and compared terraform files before and after the change for all big accounts - no diffs but also the terrascript code 90% responsible for the output was not "directly" touched
* i had various local debug scenarios running with manually crafted a-i changes to see if changes would kick in for ResourceInventory based resource management - worked as expected
* i had some code introduced in the codepath where output was written to vault to notify me when desired and current would differ - this way i was able to confirm correctness besides tests

File: reconcile/terraform_resources.py (hunk 7 above)
Comment (steveteahan) [reply reconcile/terraform_resources.py]: I spoke to @Piojo about this earlier, but `Mapping` does in fact have `get`. When in doubt, we can always check [here](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes).

We have internal documentation on this topic [here](https://service.pages.redhat.com/dev-guidelines/docs/appsre/python-style-guide/#abstract-collection-types).

File: reconcile/utils/terraform_resource_spec.py (hunk 9 above)
Comment (steveteahan) [reply reconcile/utils/terraform_resource_spec.py]: That makes even more sense, thanks!

**Comment by steveteahan** [comment]
Thanks for the info on what was tested! Your PR summary was already very detailed, but I do think it's helpful to include some of the information about manual tests.

File: reconcile/utils/terraform_resource_spec.py [hunk 20]
```diff
@@ -0,0 +1,88 @@
+from dataclasses import dataclass, field
+import json
+from typing import Any, Optional, cast
+
+
+@dataclass
+class TerraformResourceSpec:
+
+    resource: dict[str, Any]
```
Comment (janboll) [inline reconcile/utils/terraform_resource_spec.py:12]: I prefer property declaration as done in `TerraformResourceIdentifier`. Is there a reason for the diff in implementation?

File: reconcile/terraform_resources.py [hunk 21]
```diff
@@ -549,15 +549,15 @@ def cleanup_and_exit(tf=None, status=False, working_dirs=None):
...
+def write_outputs_to_vault(vault_path: str, resoure_specs: TerraformResourceSpecDict) -> None:
     integration_name = QONTRACT_INTEGRATION.replace('_', '-')
-    vault_client = VaultClient()
-    for cluster, namespace, _, data in ri:
-        for name, d_item in data['desired'].items():
-            secret_path = \
-                f"{vault_path}/{integration_name}/{cluster}/{namespace}/{name}"
-            secret = {'path': secret_path, 'data': d_item.body['data']}
-            vault_client.write(secret)
+    vault_client = cast(_VaultClient, VaultClient())
```
Comment (janboll) [inline reconcile/terraform_resources.py:554]: What is happening here?

File: reconcile/terraform_resources.py (hunk 7 above)
Comment (geoberle) [reply reconcile/terraform_resources.py]: i tried to apply it where possible but stopped at places where it would have leaked into ares of the code that are not within the scope of this PR.

expect more typehints in upcoming PRs but not this one

File: setup.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (geoberle) [reply setup.py]: i rebased on top of @patjlm semver so this is of no concern anymore in this PR

File: reconcile/terraform_resources.py (hunk 21 above)
Comment (geoberle) [reply reconcile/terraform_resources.py:554]: this is mostly just hopping through some hoops to make mypy happy type wise

File: reconcile/utils/terraform_resource_spec.py (hunk 20 above)
Comment (geoberle) [reply reconcile/utils/terraform_resource_spec.py]: consider this the minimum viable product to represent a terraform resource spec from the [schema](https://github.com/app-sre/qontract-schemas/blob/8ebce0bf85270b09c0919e0195fb1fc2ba424f44/graphql-schemas/schema.yml#L968) . this enables us work with the rich context this dataclass provides without going all in for
* mapping out the sub dataclasses
* adapting terrascript client
this would have widened the scope of this PR, which is desired state and vault management.

this is something for upcoming refactorings for sure!

File: reconcile/utils/terraform_resource_spec.py [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+        annotation_str = self.resource.get("annotations")
+        if annotation_str:
+            return json.loads(annotation_str)
+        else:
+            return {}
+
+    def get_secret_field(self, field: str) -> Optional[str]:
+        return self.secret.get(field)
+
+    def id_object(self) -> "TerraformResourceIdentifier":
```
Comment (lechuk47) [inline reconcile/utils/terraform_resource_spec.py:54]: Consider using the  https://peps.python.org/pep-0563/ approach to solve it. It's available on python 3.7+ and it's nicer IMO. 

TLDR; Just add this import and you can set the class identifier. 

```
from __future__ import annotations
```

File: reconcile/utils/terraform_resource_spec.py [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+        return TerraformResourceIdentifier.from_dict(self.resource)
+
+    def build_oc_secret(
```
Comment (janboll) [inline reconcile/utils/terraform_resource_spec.py:57]: What do you think of moving this to OpenshiftRecource like:
```
def createFromTerraformResourcespec(cls, tr: TerraformResourceSpec):
   or = cls()
   ....
```

 Cause it holds plenty of details regarding Openshift resources.

File: reconcile/utils/terraform_client.py [hunk 24]
```diff
@@ -511,38 +449,57 @@ def format_output(self, output, type):
...
-            body["metadata"]["annotations"].update(anno_dict)
+        self._populate_terraform_output_secrets(
+            resource_specs, existing_secets, self.integration_prefix, replicas_info
+        )
 
-        for k, v in data.items():
-            if self.integration_prefix in k:
+    @staticmethod
+    def _populate_terraform_output_secrets(resource_specs: TerraformResourceSpecDict,
+                                           existing_secrets: Mapping[str, Mapping[str, Mapping[str, str]]],
```
Comment (janboll) [inline reconcile/utils/terraform_client.py:472]: See above regarding `existing_secets`

File: reconcile/terraform_resources.py (hunk 7 above)
Comment (steveteahan) [reply reconcile/terraform_resources.py]: Thanks for taking care of those updates!

**Review by steveteahan — APPROVED** [review]

File: reconcile/utils/terraform_client.py [hunk 25]
```diff
@@ -511,38 +449,57 @@ def format_output(self, output, type):
...
-        }
+    def populate_terraform_output_secrets(self, resource_specs: TerraformResourceSpecDict,
+                                          init_rds_replica_source: bool = False) -> None:
+        """
+        find the terraform output data for each resource spec and populate its `secret` field.
+        if the `init_rds_replica_source` a replica RDS gets its DB user and password fields
+        populated by looking at the replica source DB.
+        """
+        self.init_outputs()  # get updated output
+        existing_secets = self.get_terraform_output_secrets()
```
[the remark this answers is not in the document]
Comment (geoberle) [reply reconcile/utils/terraform_client.py]: renamed

File: reconcile/utils/terraform_client.py (hunk 24 above)
Comment (geoberle) [reply reconcile/utils/terraform_client.py]: renamed

**Review by maorfr — APPROVED** [review]
LGTM

huge kudos on the split to commits and the explanations. i followed your directions and it made perfect sense.

File: reconcile/terraform_resources.py [hunk 26]
```diff
@@ -491,6 +520,24 @@ def filter_tf_namespaces(namespaces, account_name):
...
+def init_tf_resource_specs(
+    namespaces: list[dict[str, Any]], account_name: Optional[str]
+) -> TerraformResourceSpecDict:
+    resource_specs: TerraformResourceSpecDict = {}
+    for namespace_info in namespaces:
+        if not namespace_info.get("managedTerraformResources"):
+            continue
+        tf_resources = namespace_info.get("terraformResources") or []
+        for resource in tf_resources:
+            if account_name is None or resource["account"] == account_name:
```
[the remark this answers is not in the document]
Comment (geoberle) [reply reconcile/terraform_resources.py:532]: the account_name comes from [cli.py](https://github.com/app-sre/qontract-reconcile/blob/6845d95897c6808424e5d9909e9f4486cf566a48/reconcile/cli.py#L382)

File: reconcile/utils/terraform_resource_spec.py [hunk 27]
```diff
@@ -0,0 +1,88 @@
...
+            return json.loads(annotation_str)
+        else:
+            return {}
+
+    def get_secret_field(self, field: str) -> Optional[str]:
+        return self.secret.get(field)
+
+
+@dataclass(frozen=True)
+class TerraformResourceIdentifier:
```
[the remark this answers is not in the document]
Comment (geoberle) [reply reconcile/utils/terraform_resource_spec.py]: yes exactly. i went for TerraformResourceUniqueKey in [d6d92ed](https://github.com/app-sre/qontract-reconcile/pull/2331/commits/d6d92ed18da1ce15f75570d539c7572f1432e23b)

File: reconcile/terraform_resources.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (maorfr) [reply reconcile/terraform_resources.py:532]: as long as it has a purpose, i'm ok with it. it is also compatible with the overall plan of supporting multiple cloud providers.

File: reconcile/utils/terraform_client.py [hunk 28]
```diff
@@ -331,120 +331,66 @@ def terraform_apply(self, apply_spec):
...
+                overrides = tf_resource.get('overrides')
+                if overrides is not None:
+                    override_replica_src_db = json.loads(overrides).get(
+                        'replicate_source_db'
+                    )
+                if override_replica_src_db is not None:
+                    replica_src_db = override_replica_src_db
+
+                # Getting whatever we probed here
+                replica_src = replica_src_db
```
[the remark this answers is not in the document]
Comment (geoberle) [reply reconcile/utils/terraform_client.py]: in the linked MR i can see the removal of the replicate_source_db field from the rds-defaults schema. i also checked if any replicate_source_db fields are currently present in any of our RDS overrides, and did not find any. i wonder what the purpose of such overrides directly in the RDS declaration would be if the the only other option to specify the source is on the RDS object directly as well.

does that render the entire code block obsolete and can rely on the replicate_source field on RDS only?

[the remark this answers is not in the document]
Comment (maorfr) [reply reconcile/utils/terraform_client.py]: yes!

File: reconcile/terraform_resources.py [hunk 29]
```diff
@@ -560,6 +560,22 @@ def write_outputs_to_vault(vault_path, ri):
             vault_client.write(secret)
 
 
+def populate_desired_state(ri: ResourceInventory, resource_specs: TerraformResourceSpecDict) -> None:
+    for spec in resource_specs.values():
+        if ri.is_cluster_present(spec.cluster_name):
```
[the remark this answers is not in the document]
Comment (geoberle) [reply reconcile/terraform_resources.py:565]: it happens when clusters are skipped during ResourceInventory initialization (e.g. disabled clusters)

File: reconcile/utils/terraform_resource_spec.py [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+    ) -> "TerraformResourceIdentifier":
+        identifier, provider = output_prefix.rsplit("-", 1)
+        return TerraformResourceIdentifier(
+            identifier=identifier,
+            provider=provider,
+            account=account,
+        )
+
+
+TerraformResourceSpecDict = dict[TerraformResourceIdentifier, TerraformResourceSpec]
```
[the remark this answers is not in the document]
Comment (geoberle) [reply reconcile/utils/terraform_resource_spec.py]: renamed to TerraformResourceSpecInventory
i'm not going to turn this into a dataclass since i have no idea yet how the interface of such an inventory would look like for other providers like GCP

File: reconcile/utils/terraform_client.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (geoberle) [reply reconcile/utils/terraform_client.py]: [f602827](https://github.com/app-sre/qontract-reconcile/pull/2331/commits/f6028273417857be0491f1949e94641d26a7d57e)

**Comment by maorfr** [comment]
https://issues.redhat.com/browse/APPSRE-5390
